### PR TITLE
s3: Add header option to PutXXX() methods

### PR DESCRIPTION
**Pull request by surma**
As requested, this is the [same PR as the one I opened on launchpad](https://code.launchpad.net/~surma/goamz/put-headers/+merge/232038).

It is supposed to implement this enhancement request: https://bugs.launchpad.net/goamz/+bug/1321254

### CODE REVIEW

File: s3/s3.go [hunk 1]
```diff
@@ -192,16 +192,22 @@ func (b *Bucket) Put(path string, data []byte, contType string, perm ACL) error
...
+	return b.PutReaderWithHeader(path, r, length, contType, perm, http.Header{})
+}
+
+// PutReader inserts an object into the S3 bucket by consuming data
+// from r until EOF. The HTTP PUT request will have the additional
+// headers provided. This is useful for cache control etc.
+func (b *Bucket) PutReaderWithHeader(path string, r io.Reader, length int64, contType string, perm ACL, hdrs http.Header) error {
+	hdrs.Set("Content-Length", strconv.FormatInt(length, 10))
+	hdrs.Set("Content-Type", contType)
+	hdrs.Set("x-amz-acl", string(perm))
```
Comment (samsalisbury) [inline s3/s3.go:204]: This line is problematic if someone wants to use these headers: x-amz-grant-read, x-amz-grant-write, x-amz-grant-full-control: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html (see **Access Permissions**)

These headers were not possible before this commit either of course...

Comment (samsalisbury) [inline s3/s3.go:204]: OK, this is a moot point, I'll open an issue & PR here instead... https://github.com/goamz/goamz/blob/master/s3/s3.go#L393

**Comment by surma** [comment]
@dimitern I added a test. However, the tests don’t run locally (they don’t run on `master`, either). Maybe you can take a quick look on what’s going on? :)

**Comment by dimitern** [comment]
@surma Thanks! The test looks good. However, can I ask you to re-target this patch for the `v3-unstable` branch, as that's where the integration work will be done? About the tests - what do you mean locally? I've just realized the `v3-unstable` branch has incorrect Travis CI config for running tests - I'll fix that now.

**Comment by surma** [comment]
@dimitern I get the following:

```
$GOPATH/src/github.com/go-amz/amz/s3 [v3-unstable] $ go test
# github.com/go-amz/amz/s3_test
./s3_test.go:36: undefined: s3.SetAttemptStrategy
./s3_test.go:45: undefined: s3.SetAttemptStrategy
./s3_test.go:278: undefined: s3.SetAttemptStrategy
./s3_test.go:279: undefined: s3.AttemptStrategy
./s3_test.go:281: undefined: s3.AttemptStrategy
./s3_test.go:283: undefined: s3.AttemptStrategy
./s3i_test.go:430: undefined: s3.SetListPartsMax
./sign_test.go:21: undefined: s3.Sign
./sign_test.go:35: undefined: s3.Sign
./sign_test.go:53: undefined: s3.Sign
./sign_test.go:53: too many errors
FAIL    github.com/go-amz/amz/s3 [build failed]
```

**Comment by dimitern** [comment]
@surma Sorry, I had to fix the v3-unstable branch. If you fetch the changes and rebase yours it should be fine.

**Comment by surma** [comment]
@dimitern Rebased to `v3-unstable`. Tests running and passing :shipit:

**Comment by dimitern** [comment]
Hey @surma, sorry I totally dropped the ball on this one :( I should've responded earlier.

Thanks for rebasing to `v3-unstable`, however please open a new PR with the changes targeted to `v3-unstable`, as this cannot be merged as-is in its original target `v1` and I'll close it.

**Pull request closed without merging**